# Rule to test the unfinished pipeline
import os
from pathlib import Path

def get_drep_genomes(wildcards):
    ckpt = checkpoints.dereplicate_bins.get(**wildcards)
    drep_dir = Path(ckpt.output.dereplicated_bins)
    return sorted(drep_dir.glob("*.fa"))

rule downstream_test:
    input:
        f"{outdir}/results/06_binning/metabat2/{{sample_pool}}/{{sample_pool}}_metabat2.done",
        f"{outdir}/results/06_binning/maxbin2/{{sample_pool}}/{{sample_pool}}_maxbin2.done",
        f"{outdir}/results/06_binning/dastool/{{sample_pool}}/{{sample_pool}}_DASTool.done",
        f"{outdir}/results/06_binning/drep/dereplicated_genomes/drep.done",
        # test_target,
        f"{outdir}/results/03_assembly/coassembly/pools/{{sample_pool}}_forward.fastq.gz",
        f"{outdir}/results/03_assembly/coassembly/assembly_{{sample_pool}}/{{sample_pool}}_assembly.fasta.gz",
        # f"{outdir}/results/03_assembly/coassembly/pools/{{sample_pool}}_rev.fastq.gz",
        f"{outdir}/results/04_gene_prediction/prodigal/{{sample_pool}}/{{sample_pool}}_orfs.fna",
        f"{outdir}/results/05_prokaryote_annotation/CAT/{{sample_pool}}/{{sample_pool}}.contig2classification.names.summarise.txt",
        f"{outdir}/results/05_prokaryote_annotation/eggnog/{{sample_pool}}/{{sample_pool}}.emapper.annotations",
        # f"{outdir}/results/05_prokaryote_annotation/MetaPhlAn/temp_MetaPhlAn/{{sample_pool}}.txt",
        # f"{outdir}/results/05_prokaryote_annotation/MetaPhlAn/merged_abundance_table.txt",
        f"{outdir}/results/03_assembly/size_filtered/{{sample_pool}}_{minsize}/contigs_{{sample_pool}}_{minsize}.fasta",
        f"{outdir}/results/04_gene_prediction/prodigal/{{sample_pool}}/{{sample_pool}}_proteins.faa",
        # f"{outdir}/results/04_gene_prediction/whokaryote/{{sample_pool}}/eukaryotes.fasta",
        # f"{outdir}/results/04_gene_prediction/augustify/{{sample_pool}}/{{sample_pool}}_eukproteins.gff",
        f"{outdir}/results/06_binning/checkm2/{{sample_pool}}/quality_report.tsv",
        # f"{outdir}/results/06_binning/drep/checkm2_genomeinfo/{{sample_pool}}_genomeinfo.tsv",
        # f"{outdir}/results/06_binning/drep/combined_genomeinfo.tsv",
        f"{outdir}/results/07_maglinkage/{{sample_pool}}/markermag/{{sample_pool}}_linkages_by_genome.txt",
        f"{outdir}/results/06_binning/BAT/{{sample_pool}}/{{sample_pool}}_BAT.done",
        f"{outdir}/results/08_plots/{{sample_pool}}/{{sample_pool}}_krona.html",
        f"{outdir}/results/08_plots/{{sample_pool}}/{{sample_pool}}_bins_scatterplot.html",
        # f"{outdir}/results/08_BGC/antismash/{{sample_pool}}/bacterial/index.html",
        expand(
            f"{outdir}/results/08_BGC/antismash/{{sample_pool}}/bacterial/{{genome}}/.antismash.done",
            sample_pool=samples["sample_pool"],
            genome=lambda wc: get_drep_genomes(wc)
        ),
        f"{outdir}/results/08_BGC/antismash/{{sample_pool}}/fungal/index.html",
        # f"{outdir}/results/06_binning/BAT/{{sample_pool}}/{{sample_pool}}.bin2classification.names.txt"
        #f"{outdir}/results/08_plots/{{sample_pool}}/{{sample_pool}}_bins_scatterplot.html",
        f"{outdir}/results/08_plots/Orochi_report.html"


    output:
#        test_file=f"{outdir}/results/05_test/{{sample}}/{{sample}}_test.txt",
        test_file1=f"{outdir}/results/05_test/{{sample_pool}}/{{sample_pool}}_test.txt"
    run:
        shell("echo {input}")
#        shell("touch {output.test_file} && touch {output.test_file1}")
        shell("touch {output.test_file1}")

        # shell("touch {output.test_file}")
